components/style/gecko/data.rs

```diff
@@ -113,30 +109,29 @@ impl PerDocumentStyleData {
 }
 
 impl PerDocumentStyleDataImpl {
-    /// Recreate the style data if the stylesheets have changed.
-    pub fn flush_stylesheets(&mut self) {
-        let mut stylist = if self.device_changed || self.stylesheets_changed {
-            Some(Arc::get_mut(&mut self.stylist).unwrap())
-        } else {
-            None
-        };
-
-        if self.device_changed {
-            Arc::get_mut(&mut stylist.as_mut().unwrap().device).unwrap().reset();
-            self.device_changed = false;
-            // Force a stylesheet flush if the device has changed.
-            self.stylesheets_changed = true;
+    /// Reset the device state because it may have changed.
+    ///
+    /// Implies also a stylesheet flush.
+    pub fn reset_device(&mut self) {
+        {
+            let mut stylist = Arc::get_mut(&mut self.stylist).unwrap();
+            Arc::get_mut(&mut stylist.device).unwrap().reset();
         }
+        self.stylesheets_changed = true;
```

Comment:
Oh, and just so that I understand - if Gecko does provide us with such a restyle hint, will it avoid invoke RebuildAllStyleData? Because otherwise we're still going to dirty everything.

I guess I'm probably enough context here that it makes sense for heycam to review it, since he reviewed the media query stuff.